src/app/src/main/kotlin/org/vaccineimpact/orderlyweb/db/AuthorizationRepository.kt

```diff
@@ -140,6 +141,37 @@ class OrderlyAuthorizationRepository : AuthorizationRepository
 
     }
 
+    override fun getReportReaders(reportName: String): Map<User, Scope>
+    {
+        //Returns all users which can read the report, along with their report read scope (global or report-specific)
+        JooqContext().use {
+            val result = it.dsl.select(ORDERLYWEB_USER.USERNAME, ORDERLYWEB_USER.DISPLAY_NAME, ORDERLYWEB_USER.EMAIL,
+                    ORDERLYWEB_USER.USER_SOURCE, ORDERLYWEB_USER.LAST_LOGGED_IN,
+                    ORDERLYWEB_USER_GROUP_GLOBAL_PERMISSION.ID, ORDERLYWEB_USER_GROUP_REPORT_PERMISSION.ID)
+                    .from(ORDERLYWEB_USER)
+                    .join(ORDERLYWEB_USER_GROUP)
+                    .on(ORDERLYWEB_USER.EMAIL.eq(ORDERLYWEB_USER_GROUP.ID))
+                    .join(ORDERLYWEB_USER_GROUP_PERMISSION)
+                    .on(ORDERLYWEB_USER_GROUP_PERMISSION.USER_GROUP.eq(ORDERLYWEB_USER_GROUP.ID))
+                    .leftJoin(ORDERLYWEB_USER_GROUP_GLOBAL_PERMISSION)
+                    .on(ORDERLYWEB_USER_GROUP_PERMISSION.ID.eq(ORDERLYWEB_USER_GROUP_GLOBAL_PERMISSION.ID))
+                    .leftJoin(ORDERLYWEB_USER_GROUP_REPORT_PERMISSION)
+                    .on(ORDERLYWEB_USER_GROUP_PERMISSION.ID.eq(ORDERLYWEB_USER_GROUP_REPORT_PERMISSION.ID))
+                    .where(ORDERLYWEB_USER_GROUP_PERMISSION.PERMISSION.eq("reports.read"))
+                    .and(ORDERLYWEB_USER_GROUP_GLOBAL_PERMISSION.ID.isNotNull.or(ORDERLYWEB_USER_GROUP_REPORT_PERMISSION.REPORT.eq(reportName)))
+                    .fetch()
+
+            return result.map{
+                it.into(User::class.java) to
+                        if (it[ORDERLYWEB_USER_GROUP_GLOBAL_PERMISSION.ID] != null)
+                            Scope.Global()
+                        else
+                            Scope.Specific("report", reportName)
+            }.toMap()
```

Comment:
I think `map` followed by `toMap` could be replaced by `associateBy` https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.sequences/associate-by.html